### DISCUSSION TITLE
Why is the access token not saved in the Cookie?

### DISCUSSION THREAD

**Discussion by diegoleme** [directus/directus#10103] [Archive]
What if the directus API accepted one more authentication mode via Cookie, in addition to Header Authorization and Query string.

And instead of saving only the `refresh_token` in cookie, save the `access_token` too?

It would be totally transparent for customers who were going to use the authentication.

**Comment by eikaramba** [comment]
i have the same question. given that saving the acces_token in local storage is not recommended anymore, i would argue that acces_token, refresh_token and expiration get each returned as their own cookie. expiration and refresh_token with longer TTL. The expiration cookie has the expiration timestamp of the acces_token cookie. by doing this one could easily refresh the acces_token without needing to make any api call before (you exactly know whether it is already invalid or not and can directly refresh it instead of handling an error)

This is currently blocking us unfortunately from using directus as we don't want to add another middlelayer inbetween just for the cookie handling.

**Reply by rijkvanzanten** [reply]
Not sure if I'm following what you mean here:

> by doing this one could easily refresh the acces_token without needing to make any api call before (you exactly know whether it is already invalid or not and can directly refresh it instead of handling an error)

How would you know whether or not the cookie is valid or not? Directus' cookies can't be read from JS, so you won't be able to know whether or not the cookie is valid (or even whether or not the cookie exists in the first place) without making an API call.

**Reply by eikaramba** [reply]
thanks for getting back. i am reading the cookie from server side. you are right that evidently i would not be able to say whether the access_token is valid, but is that really a problem? i mean there are only a certain number of scenarious:
a) cookie not readable, either because it is expired or not set -> look at refresh_token and refresh or if not set as well, then user is not logged in
b) cookie is set and acces_token can be read (is at least a readable JWT token). then i would just continue and use that for authentication of any api calls. the actual validation happens as soon as the first api call is made with that cookie and fails.

of course there is no reason why one would not be able to also try to get the current user on every server request and then if that fails handle it accordingly.

btw i am using sveltekit with SSR. meaning there is a hook which runs on server for every public request.

**Reply by diegoleme** [reply]
Exactly! 

It could have a flow that is completely transparent to a web application. 

JS cannot read the access token and refresh token increases application security by preventing XSS attacks.

**Reply by t7tran** [reply]
Just wanted to throw in my comments on this topic. We need to serve content within the admin app's custom module/interface/display. If content references an image from the file library, it is blocked due to no access token (#9062). Having access token in the cookies can be a fallback for the API to read.

My plan is to have access token injected into a secure cookie whenever it is issued/refreshed. Then the cookie can be read if no access token presents in the header/query string. Hopefully, it can be achieved via extension. But it would be great to see this support built in.

**Reply by rijkvanzanten** [reply]
@t7tran also keep an eye on https://github.com/directus/directus/pull/10592, which adds a v-image component that uses the header for auth 👍🏻

**Reply by t7tran** [reply]
Thanks mate. While it is definitely a good idea, I guess it only supports Vue-based content. We need to render static HTML within a web-component, or even HTML-binding to a `div`.

**Comment by b-yanis** [comment]
so today we agree that if we use 'cookie' mode, the only solution is to store the access token in loacalstorage (or session storage) ?
And also to manage to refresh of the token manually ?

**Reply by rijkvanzanten** [reply]
https://docs.directus.io/reference/authentication.html should have all the info 👍🏻

**Reply by eikaramba** [reply]
that is not true, you can just store the access_token yourself in a cookie. i do this see https://docs.directus.io/blog/directus-auth-sveltekit.html

**Reply by rijkvanzanten** [reply]
Or use `session` mode so it's handled all in a cookie automatically
